Source/Core/Core/HW/HSP/HSP.cpp

```diff
@@ -0,0 +1,67 @@
+// Copyright 2018 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "Core/HW/HSP/HSP.h"
+
+#include <memory>
+
+#include "Common/Logging/Log.h"
+#include "Core/Config/MainSettings.h"
+#include "Core/ConfigManager.h"
+#include "Core/HW/HSP/HSP_Device.h"
+
+namespace HSP
+{
+static std::unique_ptr<IHSPDevice> s_device;
+
+void Init()
+{
+  AddDevice(Config::Get(Config::MAIN_HSP_DEVICE));
+}
+
+void Shutdown()
+{
+  RemoveDevice();
+}
+
+u64 Read(u32 address)
+{
+  DEBUG_LOG_FMT(HSP, "HSP read from 0x{:08x}", address);
+  if (s_device)
+    return s_device->Read(address);
+  return 0;
+}
+
+void Write(u32 address, u64 value)
+{
+  DEBUG_LOG_FMT(HSP, "HSP write to 0x{:08x}: 0x{:016x}", address, value);
+  if (s_device)
+    s_device->Write(address, value);
+}
+
+void DoState(PointerWrap& p)
+{
+  if (s_device)
+    s_device->DoState(p);
+}
+
+void AddDevice(std::unique_ptr<IHSPDevice> device)
+{
+  // Delete the old device
+  RemoveDevice();
+
+  // Set the new one
+  s_device = std::move(device);
```

Comment:
The call to `RemoveDevice()` is unneeded, as assigning to a `unique_ptr` deletes the old value (if present), as if `reset()` were called.